### ISSUE TITLE
Recursive config loading does not load everything

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1031]
In GitLab by @abaer-kit on May 3, 2023, 17:46

Hello @dbeltrankyl and @kinow,

### Autosubmit Version

v4.0.78 (local installation)

### Which task has issues? Where is the log(If applicable)

autosubmit create

### Summary

After the fixes mentioned in #1026, I found some weird behavior with the includes. Basically, this is an extension to [this comment](https://earth.bsc.es/gitlab/es/autosubmit/-/issues/1026#note_213319).
For the following includes (see below for the full configuration)

`include.yml`:
```
DEFAULT:
  CUSTOM_CONFIG: 
    PRE:
      - "%PROJDIR%/config/common"
      - "%PROJDIR%/config/testfile.yml"
      - "%PROJDIR%/config/testfile2.yml"
      - "%PROJDIR%/config/testfile-%DEFAULT.HPCARCH%.yml"
    POST:
      - "/home/andreas/autosubmit/config/platforms_user.yml"
```
I realized the following:
1. Keys already defined by files in `%PROJDIR%/config/common` will not be overwritten if they are again defined (e.g. in `testfile.yml`).
2. Files containing placeholders (e.g. `testfile-%DEFAULT.HPCARCH%.yml`) will not be loaded, even though they appear with `autosubmit report` in the DEFAULT.CUSTOM_CONFIG.PRE list.
3. 2. also applies for the POST section.

Instead if I put all the content of the PRE and POST keys inside the `minimal.yml` file, they will be processed as expected.

### Steps to reproduce

The full configuration used is as follows:
`minimal.yml`:
```
CONFIG:
  AUTOSUBMIT_VERSION: "4.0.76"
  TOTALJOBS: 20
  MAXWAITINGJOBS: 20
DEFAULT:
  EXPID: "a00d"
  HPCARCH: "levante"
  TEMPLATE: "NWP_EXT_DATA"
  #hint: use %PROJDIR% to point to the project folder (where the project is cloned)
  CUSTOM_CONFIG:
    PRE:
      - "%PROJDIR%/config/include"

PROJECT:
  PROJECT_TYPE: "git"
  PROJECT_DESTINATION: "auto-icon"
GIT:
  PROJECT_ORIGIN: "git@gitlab.dkrz.de:auto-icon/auto-icon.git"
  PROJECT_BRANCH: "refactor-config-loading"
  PROJECT_COMMIT: ""
  PROJECT_SUBMODULES: ""
  FETCH_SINGLE_BRANCH: True
```
`%PROJDIR%/config/common/expdef.yml`:
```
EXPERIMENT:
  # Standard autosubmit experiment directives
  DATELIST: 20000101
  MEMBERS: "default"
  CHUNKSIZEUNIT: month
  CHUNKSIZE: 1
  NUMCHUNKS: 1
  CHUNKINI: ''
  CALENDAR: standard
NAMELIST:
  MASTER: 'icon_master.namelist'
  EXP: '%DEFAULT.TEMPLATE%.nml'
DIRECTORIES:
  # If INDIR, OUTDIR and/or REFDIR are treatet as full absolute paths.
  # Alternatively, the path can be concatenated as follows:
  #   PREFIX_INDIR/REL_IN/EXPNAME
  #   PREFIX_OUTDIR/REL_OUT/EXPNAME
  #   PREFIX_RELDIR/REL_REF/EXPNAME_ref
  # Loading the config file `custom/construct_path.yml`, constructs the paths.
  ICONDIR: '%HPCROOTDIR%/icon'
  INDIR: ''
  OUTDIR: ''
```
`%PROJDIR%/config/common/jobs.yml`:
```
JOBS:
  COPY_NAMELIST:
    FILE: "jobscripts/copy_namelist.sh"
    RUNNING: once
```
`%PROJDIR%/config/common/platforms.yml`:
```
PLATFORMS:
  LEVANTE:
    TYPE: slurm
    HOST: levante.dkrz.de
    USER: ''
    PROJECT: ''
    SCRATCH_DIR: '/work'
    MAX_WALLCLOCK: '08:00'
    PROCESSORS_PER_NODE: 128
    CUSTOM_DIRECTIVES: '#SBATCH --partition=compute'
  
```
`%PROJDIR%/config/include/include.yml`:
```
DEFAULT:
  CUSTOM_CONFIG: 
    PRE:
      - "%PROJDIR%/config/common"
      - "%PROJDIR%/config/testfile.yml"
      - "%PROJDIR%/config/testfile2.yml"
      - "%PROJDIR%/config/testfile-%DEFAULT.HPCARCH%.yml"
    POST:
      - "/home/andreas/autosubmit/config/platforms_user.yml"
```

`testfile.yml`:
```
TESTKEY: 'abcd'
DIRECTORIES:
  INDIR: 'my-updated-indir'
  TESTDIR: 'another-dir'
```

`testfile2.yml`:
```
TESTKEY-TWO: 'HPCARCH is %DEFAULT.HPCARCH%'
```

`testfile-levante.yml`:
```
TESTKEY-LEVANTE: 'L-abcd'
```

`/home/andreas/autosubmit/config/platforms_user.yml`:
```
Platforms:
  levante-login:
    USER: abcd
    PROJECT: efgh
  levante:
    USER: abcd
    PROJECT: efgh
...
```

### What is the current bug behavior?

According to autosubmit report, the loaded config from the testfiles is as follows (full report [a00d_parameter_list_20230503-172319-mod.txt](https://earth.bsc.es/gitlab/es/autosubmit/uploads/60cfb3aaea955b79c2096d814d4a8bb1/a00d_parameter_list_20230503-172319-mod.txt)):
```
...
DEFAULT.CUSTOM_CONFIG.PRE=/home/andreas/research/experiments/a00d/proj/auto-icon/config/common,/home/andreas/research/experiments/a00d/proj/auto-icon/config/testfile.yml,/home/andreas/research/experiments/a00d/proj/auto-icon/config/testfile2.yml,/home/andreas/research/experiments/a00d/proj/auto-icon/config/testfile-levante.yml
DEFAULT.CUSTOM_CONFIG.POST=/home/andreas/autosubmit/config/platforms_user.yml
...
DIRECTORIES.INDIR=-
DIRECTORIES.OUTDIR=-
DIRECTORIES.TESTDIR=another-dir
TESTKEY=abcd
TESTKEY-TWO=HPCARCH is levante
...
PLATFORMS.LEVANTE-LOGIN.USER=abcd
PLATFORMS.LEVANTE-LOGIN.PROJECT=abcd
PLATFORMS.LEVANTE.USER=-
PLATFORMS.LEVANTE.PROJECT=-
...
```
This misses `TESTKEY-LEVANTE=L-abcd` from `testfile-levante.yml`.
Furthermore, `DIRECTORIES.INDIR` is empty (as specified in `expdef.yml`) not equal to 'my-updated-indir' as specified in the later-loaded `testfile.yml`. The same applies to `PLATFORMS.LEVANTE.USER/PROJECT` that should have been overwritten by the file in the POST section.

### What is the expected correct behavior?

If all contents from `include.yml` are inserted into `minimal.yml` (and not loading `include.yml`), the loaded config from the testfiles is as follows (full report [a00d_parameter_list_20230503-173621-mod.txt](https://earth.bsc.es/gitlab/es/autosubmit/uploads/dddca47db480a01e5a07c1ef8b48b4aa/a00d_parameter_list_20230503-173621-mod.txt)):
```
...
DEFAULT.CUSTOM_CONFIG.PRE=/home/andreas/research/experiments/a00d/proj/auto-icon/config/common,/home/andreas/research/experiments/a00d/proj/auto-icon/config/testfile.yml,/home/andreas/research/experiments/a00d/proj/auto-icon/config/testfile2.yml,/home/andreas/research/experiments/a00d/proj/auto-icon/config/testfile-levante.yml
DEFAULT.CUSTOM_CONFIG.POST=/home/andreas/autosubmit/config/platforms_user.yml
...
DIRECTORIES.INDIR=my-updated-indir
DIRECTORIES.OUTDIR=-
DIRECTORIES.TESTDIR=another-dir
TESTKEY=abcd
TESTKEY-TWO=HPCARCH is levante
TESTKEY-LEVANTE=L-abcd
...
PLATFORMS.LEVANTE-LOGIN.USER=abcd
PLATFORMS.LEVANTE-LOGIN.PROJECT=abcd
PLATFORMS.LEVANTE.USER=abcd
PLATFORMS.LEVANTE.PROJECT=abcd
...
```

Hereby, the order of the reported entries also differs, but I don't think this actually is an issue.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 4, 2023, 15:19

Hello, I think I solved most of the issues,

Thanks for providing the configuration. 

About the placeholders, It only affected the DEFAULT section as it was defined in the minimal. I changed a bit the loading logic so I can load all parameters specified in Minimal without affecting the load. 

About the load order, some changes would not be saved after being loaded when working with multiple files. This should happen before because I did not change that part in the last commit. Anyway, it seems solved in my local branch and your configuration.

 I will test a more complex configuration to see if there are further issues and try to update the pip tomorrow since I'll have more time.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 4, 2023, 15:37

mentioned in commit ces/autosubmit4-config-parser@75ecd06d0e4e9683cd65bf0a47941d53b2eb2ccf

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 8, 2023, 12:21

Hello, 

I did a deep test and found some issues using more complex configurations. 

I created a recursion test to help me debug it, and after a few fixes now everything I tested reported OK. Also, the performance of the function was improved.

 Likely, I'll update the tag this afternoon, I'll update this issue when avaliable

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 8, 2023, 16:28

I fixed another bug where if you put additional data in the include.yml, that is the one loading the files. It was not overwriting the key properly.

Did not find more issues. Updated in pip(4.0.79) and lumi(4.0.0b17)

Edit: 

Improved the regression test. Now I use two experiments that load similar data, but one using the minimal.yml and the other using the include.yml
